extensions/parquet/table/src/main/java/io/deephaven/parquet/table/location/ParquetTableLocation.java

```diff
@@ -848,6 +902,143 @@ private PushdownResult pushdownRowGroupMetadata(
+                        // Include one extra slot for NULL_LONG as a matching key if the filter includes nulls.
+                        keyMatchArray = new long[keyMatch.size() + 1];
+                        keyMatchArray[keyMatch.size()] = NULL_LONG;
+                    } else if (filterNullBehavior == BasePushdownFilterContext.FilterNullBehavior.EXCLUDES_NULLS) {
+                        keyMatchArray = new long[keyMatch.size()];
+                    } else {
+                        throw new IllegalStateException("Unexpected null behavior: " + filterNullBehavior);
+                    }
+                    keyMatch.copyToTypedArray(0, keyMatchArray, 0, keyMatch.size());
+                }
+
+                // Make a MatchFilter with the matching dictionary key IDs. This will accept any encoded value whose
+                // dictionary index is in keyMatchArray
+                final ChunkFilter matchChunkFilter = LongChunkMatchFilterFactory.makeFilter(false, keyMatchArray);
+
+                // Now we need to apply this filter to the encoded values in the row group. We can do this by
+                // iterating the "maybe" rows in chunks, getting the encoded values for those rows, and applying the
+                // filter to those encoded values.
+                // We must shift the row keys to be relative to the row group because the value store is relative to the
+                // row group.
+                final long subRegionFirstKey = getSubRegionFirstKey(rgIdx);
+                final int CHUNK_SIZE = 4096;
+                try (final RowSet tmpRowSet = rs.asRowSet().shift(-subRegionFirstKey);
+                        final RowSequence.Iterator tmpIt = tmpRowSet.getRowSequenceIterator();
+                        final ChunkSource.GetContext getContext = valueStore.makeGetContext(CHUNK_SIZE);
+                        final WritableLongChunk<OrderedRowKeys> results =
+                                WritableLongChunk.makeWritableChunk(CHUNK_SIZE)) {
+                    while (tmpIt.hasMore()) {
+                        final RowSequence tmpRs = tmpIt.getNextRowSequenceWithLength(CHUNK_SIZE);
+                        final Chunk<? extends DictionaryKeys> valueChunk = valueStore.getChunk(getContext, tmpRs);
+                        matchChunkFilter.filter(valueChunk, tmpRs.asRowKeyChunk(), results);
+                        if (results.size() > 0) {
+                            final LongChunkIterator longIt = new LongChunkIterator(results);
+                            longIt.forEachRemaining((LongConsumer) rowKey -> {
+                                // Convert the row key to the original row key in the table.
+                                final long originalRowKey = subRegionFirstKey + rowKey;
+                                matchBuilder.appendKey(originalRowKey);
+                            });
+                        }
```

Comment:
It seems a little bit extraneous to wrap in an iterator; we should just be able to do a simple for-loop against results, no need to special case for size, right?